sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuKeyBatchingIterator.scala

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+
+import ai.rapids.cudf.{ColumnVector, NvtxColor, Table}
+import com.nvidia.spark.rapids.RapidsBuffer.SpillCallback
+
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.expressions.{Attribute, SortOrder}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Given a stream of data that is sorted by a set of keys, split the data so each batch output
+ * contains all of the keys for a given key set. This tries to get the batch sizes close to the
+ * target size. It assumes that the input batches will already be close to that size and does
+ * not try to split them too much further.
+ */
+class GpuKeyBatchingIterator private (
+    iter: Iterator[ColumnarBatch],
+    sorter: GpuSorter,
+    types: Array[DataType],
+    targetSizeBytes: Long,
+    numInputRows: GpuMetric,
+    numInputBatches: GpuMetric,
+    numOutputRows: GpuMetric,
+    numOutputBatches: GpuMetric,
+    collectTime: GpuMetric,
+    concatTime: GpuMetric,
+    totalTime: GpuMetric,
+    peakDevMemory: GpuMetric,
+    spillCallback: SpillCallback)
+    extends Iterator[ColumnarBatch] with Arm {
+  private val pending = mutable.Queue[SpillableColumnarBatch]()
+  private var pendingSize: Long = 0
+
+  TaskContext.get().addTaskCompletionListener[Unit](_ => close())
+
+  def close(): Unit = {
+    pending.foreach(_.close())
+    pending.clear()
+    pendingSize = 0
+  }
+
+  override def hasNext: Boolean = pending.nonEmpty || iter.hasNext
+
+  private def getKeyCutoff(cb: ColumnarBatch): Int = {
+    val candidates = withResource(sorter.appendProjectedColumns(cb)) { appended =>
+      withResource(GpuColumnVector.from(appended)) { table =>
+        // With data skew we don't want to include too much more in a batch than is necessary
+        // so we are going to search up to 16 evenly spaced locations in the batch, including the
+        // last entry. From that we will see what cutoff works and is closest to the size we want.
+        // The 16 is an arbitrary number. It was picked because hopefully it is small enough that
+        // the indexes will all fit in the CPU cache at once so we can process the data in an
+        // efficient way.
+        val maxRow = cb.numRows() - 1
```

Comment:
Nit: Would be nice to have a `require(maxRow >=0)` to protect in case someone later refactors and forgets to avoid passing empty batches to this function.